components/director/pkg/operation/directive.go

```diff
@@ -148,10 +151,53 @@ func (d *directive) HandleOperation(ctx context.Context, _ interface{}, next gql
 		log.C(ctx).WithError(err).Errorf("An error occurred while closing database transaction: %s", err.Error())
 		return nil, apperrors.NewInternalError("Unable to finalize database operation")
 	}
+	committed = true
 
 	return resp, nil
 }
 
+func (d *directive) concurrencyCheck(ctx context.Context, op graphql.OperationType, resCtx *gqlgen.ResolverContext, idField *string) error {
+	if op == graphql.OperationTypeCreate {
+		return nil
+	}
+
+	if idField == nil {
+		return apperrors.NewInternalError("idField from context should not be empty")
+	}
+
+	resourceID, ok := resCtx.Args[*idField].(string)
+	if !ok {
+		return apperrors.NewInternalError(fmt.Sprintf("could not get idField: %q from request context", *idField))
+	}
+
+	tenant, err := d.tenantLoaderFunc(ctx)
+	if err != nil {
+		return apperrors.NewTenantRequiredError()
+	}
+
+	app, err := d.resourceFetcherFunc(ctx, tenant, resourceID)
+	if err != nil {
+		if apperrors.IsNotFoundError(err) {
+			return err
+		}
+
+		return apperrors.NewInternalError("failed to fetch resource with id %s", resourceID)
+	}
+
+	if app.GetDeletedAt().IsZero() && app.GetUpdatedAt().IsZero() && !app.GetReady() && (app.GetError() == nil || *app.GetError() == "") { // CREATING
```

Comment:
(app.GetError() == nil || *app.GetError() == "") can be extracted as helper function as I can see this check on several places